[work in progress] Updates to email template

anniesullie opened the pull request:
* Send individual emails to recipients instead of a single mail with everyone on to: line
 * Add logos in header
 * Remove &lt;pre&gt; tags around email body so font is not monospace
 * Add a button for each jurisdiction link
 * Add a footer with contact info

Sample output HTML: https://drive.google.com/file/d/17_e0Gx86lDfYlWagsyG1qwFd_6zkCMX2/view?usp=sharing

Screenshot of current progress: 
<img width="759" alt="screen shot 2018-10-18 at 10 48 56 pm" src="https://user-images.githubusercontent.com/4331931/47195046-0fadc880-d328-11e8-87a7-e4da08b49f76.png">

Issues I ran into:
  * I don't have the right WorkElections logo because it's not uploaded anywhere, and the FairElections logo I used is not from a CDN (from the workelections.com site). I think it would be best if both images were uploaded to a CDN, but not sure how to make that happen
 * The urls in my html/text emails are broken since settings.SURVEY_MONKEY_URL_ is a dummy url with no '/' at the end. I assume it's set right in production?
 * I don't have a local mailserver, so I don't have a great way of testing that the html in my emails doesn't get mangled when it's sent and received by a mail client. Any ideas?

anniesullie commented on apps/mailman/mailer.py:
```diff
@@ -74,12 +84,16 @@ def send(self):
+            messages.append(message)
         try:
-            msg.send()
+            connection.send_messages(messages)
```
This sends each message individually. This is different than bcc, but it keeps the to: line to a single address, and was easier to test locally without a mail server.

nayelipelayo replied:
Hi Annie,

-Both logos are located here: https://www.fairelectionscenter.org/, but the work elections one is a bit small. Would uploading that one on a hidden page on that site work to get the right size logo online?

-The Survey Monkey URLs aren't dummy, but I did use the same one over and over. The URLs are currently auto-populated using our survey email creator but have been set up on the back end by @devseedgit 
We would like the template email to be receptive to having it work the same way. Selecting a jurisdiction and having the survey link populate on the email, but instead of the link on its own, have a button populate, if possible. 

-I'm not sure I understand the local mail server question. In the past with different email platforms I've been able to send a test in plain text and html. Checking it on my email and phone for formatting and other issues. I realize that different email servers will look differently (Outlook vs Gmail), I'm not sure what the response to this should be.

-Would sending each message individually make us have to send messages one by one? We wanted to have the option to do a bulk email selecting the different recipients all at once so we can save time in not having to make them one by one.

anniesullie replied:
> -Both logos are located here: https://www.fairelectionscenter.org/, but the work elections one is a bit small. Would uploading that one on a hidden page on that site work to get the right size logo online

I think it would work, but I'm not super familiar with writing these types of mails.  I worry that if there are a lot (thousands) of recipients, it can stress the server the images are hosted on. I assumed that's why the original email had an image on a CDN (https://s3.amazonaws.com/voteworker/workelections.png). I think it would be best to put both images there, but I don't know how they got there. @devseedgit do you know?

> -The Survey Monkey URLs aren't dummy, but I did use the same one over and over. The URLs are currently auto-populated using our survey email creator but have been set up on the back end by @devseedgit
We would like the template email to be receptive to having it work the same way. Selecting a jurisdiction and having the survey link populate on the email, but instead of the link on its own, have a button populate, if possible.

Sorry for the confusion here! I'm still using the same code to generate the survey URLs, just with buttons instead of links. All the buttons in the screenshot were auto-generated from the list of jurisdictions I selected in the UI. But on a local server, the code produces URLs like http://example.com734, because the local server is configured to use 'http://example.com'. Just wanted to check that it's configured correctly in production, I guess @devseedgit is the one to ask?

> -I'm not sure I understand the local mail server question. In the past with different email platforms I've been able to send a test in plain text and html. Checking it on my email and phone for formatting and other issues. I realize that different email servers will look differently (Outlook vs Gmail), I'm not sure what the response to this should be.

@devseedgit any best practices for testing the emails generated for correct CSS in Gmail and Outlook? Right now I've got raw mail printed to the console, and no local mailserver.

> -Would sending each message individually make us have to send messages one by one? We wanted to have the option to do a bulk email selecting the different recipients all at once so we can save time in not having to make them one by one.

This still takes the full list of emails it did before, but now if you list 3 emails, it sends 3 emails, each with one of the addresses on the to: line. Before, it sent one email with 3 addresses on the to: line.

nayelipelayo replied:
Ok, it sounds like we're on the page here. I'll standby for what @devseedgit @alyssadelaine @scisco say regarding q's address to them

-CND image upload on the amazon location
-local email server

alyssadelaine replied:
@anniesullie I think you should be able to use the django [backend mail server](https://docs.djangoproject.com/en/2.1/topics/email/) to test sending the e-mails, and the SurveyMonkey URL is configured correctly in production, and you can test it locally by setting it as a local environment variable; in that case, it will overwrite `example.com`

anniesullie commented on apps/mailman/templates/mailman/survey_email_html.py:
```diff
@@ -6,63 +6,42 @@ def write_html(email_text, link_html):
-    <tr>
+    <tbody><tr>
+       <td align="center" style="padding-left:30px; padding-right:30px"><img width="200" alt="WorkElections" src="https://s3.amazonaws.com/voteworker/workelections.png"></td>
+       <td align="center" style="padding-left:30px; padding-right:30px"><img width="240" alt="Fair Elections Center" src="https://www.workelections.com/graphics/meta/FairElectionsCenterLogo.png"></td>
```
These are the image urls that should be served from CDN.

nayelipelayo replied:
Would Google Drive be a good option for this?

anniesullie replied:
Looks like they used to support this, but it was turned down: https://stackoverflow.com/questions/10311092/displaying-files-e-g-images-stored-in-google-drive-on-a-website

anniesullie commented on apps/mailman/templates/mailman/survey_email_html.py:
```diff
@@ -6,63 +6,42 @@ def write_html(email_text, link_html):
+def write_button(url, label):
+    return '''
+<td style="padding:10px"><div><!--[if mso]>
+  <v:roundrect xmlns:v="urn:schemas-microsoft-com:vml" xmlns:w="urn:schemas-microsoft-com:office:word" href="http://" style="height:42px;v-text-anchor:middle;width:130px;" arcsize="10%" stroke="f" fillcolor="#3AAEE0">
```
The VML is to support desktop microsoft outlook. Button was generated from https://buttons.cm/

anniesullie commented on apps/jurisdiction/models.py:
```diff
@@ -85,7 +86,7 @@ class SurveyEmail(models.Model):
     name = models.CharField('Email label', max_length=250)
     recipients = models.TextField('List of emails', help_text='Use commas, semicolons or line breaks to separate emails. Do not enter e-mail addresses containing those special characters.')
-    email_text = models.TextField('E-mail text', help_text='This text will be displayed above the survey links.', default = 'Thank you for your participation in our survey for WorkElections.com. Please click on the link corresponding with the jurisdiction for which you would like to update information.')
+    email_text = HTMLField('E-mail text', help_text='This text will be displayed above the survey links.', default = 'Thank you for your participation in our survey for WorkElections.com. Please click on the link corresponding with the jurisdiction for which you would like to update information.')
```
@nayelipelayo: I had a little more time today, so I updated the "email text" field to be a rich text editor, so you can add/preview links.

Here is what the editor looks like:
<img width="1169" alt="screen shot 2018-10-19 at 4 30 00 pm" src="https://user-images.githubusercontent.com/4331931/47242423-b2178b80-d3bc-11e8-8f08-975b1918aa85.png">

And here is a screenshot of the email created with the fields in that editor:
<img width="715" alt="screen shot 2018-10-19 at 4 31 17 pm" src="https://user-images.githubusercontent.com/4331931/47242452-c9ef0f80-d3bc-11e8-8274-1acba50a50ff.png">

nayelipelayo commented:
@anniesullie <https://github.com/anniesullie> this looks great!

 .@alyssadelaine would you be able to help us upload the new logo onto the
amazon CDN?


On Fri, Oct 19, 2018 at 4:45 PM Alyssa Harris <notifications@github.com>
wrote:

> *@alyssadelaine* commented on this pull request.
> ------------------------------
>
> In apps/mailman/mailer.py
> <https://github.com/developmentseed/api.work.vote/pull/71#discussion_r226776866>
> :
>
> >          try:
> -            msg.send()
> +            connection.send_messages(messages)
>
> @anniesullie <https://github.com/anniesullie> I think you should be able
> to use the django backend mail server
> <https://docs.djangoproject.com/en/2.1/topics/email/> to test sending the
> e-mails, and the SurveyMonkey URL is configured correctly in production,
> and you can test it locally by setting it as a local environment variable;
> in that case, it will overwrite example.com
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/developmentseed/api.work.vote/pull/71#discussion_r226776866>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ApaOyMpyNhKVC55f4gf__lvLGA2Q-bQUks5umjnjgaJpZM4XvlLi>
> .
>


-- 
Nayeli Pelayo
Outreach Manager
Fair Elections Center
1825 K Street NW, Suite 450
<https://maps.google.com/?q=1825+K+Street+NW,+Suite+450+Washington,+D.C.+20006&entry=gmail&source=g>
Washington, D.C. 20006
<https://maps.google.com/?q=1825+K+Street+NW,+Suite+450+Washington,+D.C.+20006&entry=gmail&source=g>
Phone: (202) 248-5351
npelayo@fairelectionscenter.org
www.fairelectionscenter.org

alyssadelaine commented:
@nayelipelayo Sure, just send me (or attach here) the images which should be uploaded (or replaced) and what they should be named

nayelipelayo commented:
Tried to attach on the email thread but I don't think I was able to. This is the Work Elections one, it just needs to be added.

![work elections](https://user-images.githubusercontent.com/43421384/47243235-631f2580-d3bf-11e8-9152-22965e1c3d2a.png)

anniesullie commented:
Also the image at https://www.workelections.com/graphics/meta/FairElectionsCenterLogo.png needs to be uploaded to CDN.

nayelipelayo commented:
Can we use this for that one?

https://static.wixstatic.com/media/85cfb4_63690cf1df094663953d17a337ff011a~mv2.png/v1/fill/w_316,h_123,al_c,q_80,usm_0.66_1.00_0.01/85cfb4_63690cf1df094663953d17a337ff011a~mv2.webp


On Fri, Oct 19, 2018 at 5:15 PM anniesullie <notifications@github.com>
wrote:

> Also the image at
> https://www.workelections.com/graphics/meta/FairElectionsCenterLogo.png
> needs to be uploaded to CDN.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/developmentseed/api.work.vote/pull/71#issuecomment-431501315>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ApaOyLpHwwRxYR_fXyqnT7zKX5LUcDCcks5umkEEgaJpZM4XvlLi>
> .
>


-- 
Nayeli Pelayo
Outreach Manager
Fair Elections Center
1825 K Street NW, Suite 450
<https://maps.google.com/?q=1825+K+Street+NW,+Suite+450+Washington,+D.C.+20006&entry=gmail&source=g>
Washington, D.C. 20006
<https://maps.google.com/?q=1825+K+Street+NW,+Suite+450+Washington,+D.C.+20006&entry=gmail&source=g>
Phone: (202) 248-5351
npelayo@fairelectionscenter.org
www.fairelectionscenter.org

anniesullie commented:
@alyssadelaine Any update on getting the images in the CDN, or reviewing the pull request?

alyssadelaine commented:
@anniesullie Here are the links on S3 for the new images:

https://s3.amazonaws.com/voteworker/workelections_logo.png
https://s3.amazonaws.com/voteworker/FEC_logo.webp

And sorry, I never got the notification that my review was requested, but I can do that once this is ready. Could you let me know when you've updated the links to the images above?

anniesullie commented:
> @anniesullie Here are the links on S3 for the new images:
> 
> https://s3.amazonaws.com/voteworker/workelections_logo.png
> https://s3.amazonaws.com/voteworker/FEC_logo.webp

I updated to use these urls, but I get Access Denied errors accessing them. Is there something that needs to be done to make them public?

> And sorry, I never got the notification that my review was requested, but I can do that once this is ready. Could you let me know when you've updated the links to the images above?

Thanks! Updated.

alyssadelaine commented:
> I updated to use these urls, but I get Access Denied errors accessing them. Is there something that needs to be done to make them public?

Yes, thanks for checking! That should be fixed now

anniesullie commented:
> > I updated to use these urls, but I get Access Denied errors accessing them. Is there something that needs to be done to make them public?
> 
> Yes, thanks for checking! That should be fixed now

Verified.

@nayelipelayo, here is a screenshot of the email with the new logos:
<img width="723" alt="screen shot 2018-10-23 at 5 09 56 pm" src="https://user-images.githubusercontent.com/4331931/47390957-97a81f80-d6e6-11e8-9192-88015cc8e029.png">

alyssadelaine commented:
<img width="934" alt="screen shot 2018-10-24 at 4 14 51 pm" src="https://user-images.githubusercontent.com/6155420/47458638-0ac59a80-d7a8-11e8-953f-06386d220cb1.png">
@anniesullie @nayelipelayo Just wanted to note the new look of the e-mail header (at least on Gmail)

anniesullie commented:
> <img alt="screen shot 2018-10-24 at 4 14 51 pm" width="934" src="https://user-images.githubusercontent.com/6155420/47458638-0ac59a80-d7a8-11e8-953f-06386d220cb1.png">
> 
> @anniesullie @nayelipelayo Just wanted to note the new look of the e-mail header (at least on Gmail)

Thanks! It looks like we need a logo with a white background? Did you test the patch locally with a local mail server to get the preview?

alyssadelaine commented:
@anniesullie Yeah, that's what I was thinking too. I think the other logo in s3 has a white background.

Right, I forgot that my local setup uses the same sendgrid e-mail server used on production; if you need to do more tests on this, it might make sense to make that a part of your setup call with @scisco

anniesullie commented:
I converted the logo to gif and added a white background, new file is here:
https://drive.google.com/file/d/1c_AimGVfoInA-8oxLTmlI7ld7BZsYs05/view?usp=sharing

@alyssadelaine Would you mind uploading to the CDN?

alyssadelaine commented:
@anniesullie I also have the one with a transparent background that we use on the site (I thought it was in s3, but I think it's in the meta folder) -- should we use that, or do you prefer the white background?

anniesullie commented:
The one in the test email *does* have a transparent background ([see screenshot](https://github.com/developmentseed/api.work.vote/pull/71#issuecomment-432421159)), it is just rendering weird in gmail for some reason. So let's use the one with the white background.
